### PR TITLE
feat: Add HSM support for JWE decryption with custom OAEP implementation

### PR DESCRIPTION
**Pull request by Jenarththan2001**
## DEPENDENCY NOTICE
**This PR depends on #25 being merged first.** After #25 merges, this branch will need to be rebased with main to pick up ServerKeystoreRetriever changes.

## Purpose

Enable HSM integration for JWE payload decryption with **full algorithm support**. SunPKCS11 does not support RSA-OAEP cipher transformations, so standard JWE libraries fail with HSM-backed keys. This PR implements a custom OAEP decryption approach that works within SunPKCS11 limitations.

**Note**: This implementation resolves the limitation documented in [docs-open-banking #1070](https://github.com/wso2/docs-open-banking/pull/1070), which stated "only RSA1_5 supported with HSM". With this custom implementation, all RSA-OAEP variants are now supported.

## Goals

- Enable JWE decryption with HSM-backed private keys for all standard algorithms
- Support RSA-OAEP, RSA-OAEP-256, RSA-OAEP-384, RSA-OAEP-512, and RSA1_5
- Support A128GCM, A192GCM, A256GCM content encryption
- Maintain backward compatibility with JKS keystores
- Provide comprehensive test coverage

### Three-Step Decryption Workflow

**SunPKCS11 Limitation**: Does not implement `CKM_RSA_PKCS_OAEP` mechanism, causing `NoSuchAlgorithmException` with standard JWE libraries.

**Solution**: Custom decryption helper that bypasses the limitation:

1. **Raw RSA decryption in HSM** - Uses `RSA/ECB/NoPadding` to decrypt Content Encryption Key (CEK). Private key never leaves hardware.

2. **OAEP unpadding in software** - Manually removes OAEP padding following RFC 3447 §7.1.2 (EME-OAEP decode with MGF1 mask generation).

3. **AES-GCM content decryption** - Uses standard JCE to decrypt JWE payload with extracted CEK.

### Implementation Details

**HsmJweDecryptionHelper.java** (NEW - 343 lines):
- Custom OAEP implementation per RFC 3447
- Supports all SHA variants (SHA-1, SHA-256, SHA-384, SHA-512)
- MGF1 mask generation function
- OAEP integrity validation (prevents padding oracle attacks)
- Handles both PKCS#1 v1.5 and OAEP padding

**JwePayloadDecryptionMediator.java** (ENHANCED):
- Runtime HSM detection by key type (P11PrivateKey)
- Routes to HsmJweDecryptionHelper for P11 keys
- Falls back to Nimbus RSADecrypter for standard keys
- No configuration-based branching

**HsmJweDecryptionHelperTest.java** (NEW - 370 lines, 20 methods):
- Parameterized tests for 15 algorithm combinations
- Unit tests for cryptographic primitives (MGF1, OAEP unpadding, hash selection)
- Edge case testing (malformed JWE, invalid padding, unsupported algorithms)

**Note on ServerKeystoreRetriever**: HSM-aware key retrieval provided by PR #25. This PR builds on that foundation.

## Release Note

**Feature**: HSM Support for JWE Payload Decryption

JWE Payload Processing Policy now supports HSM integration with custom OAEP implementation.

**Algorithm Support**:
- HSM: RSA1_5, RSA-OAEP, RSA-OAEP-256, RSA-OAEP-384, RSA-OAEP-512
- Content Encryption: A128GCM, A192GCM, A256GCM
- JKS: All algorithms supported (no change)

**Note**: Custom cryptographic implementation. Security review recommended before production deployment.

## Documentation

**Related Documentation PR**: [docs-open-banking #1070](https://github.com/wso2/docs-open-banking/pull/1070)

⚠️ **Note**: PR #1070 documents a limitation stating "only RSA1_5 supported with HSM" because RSA-OAEP variants fail with SunPKCS11. **This PR (custom OAEP implementation) removes that limitation** by implementing custom OAEP decryption. If this PR is merged, the documentation PR should be **closed/reverted** since the limitation no longer exists - all algorithms (RSA1_5, RSA-OAEP, RSA-OAEP-256, RSA-OAEP-384, RSA-OAEP-512) will be supported with HSM.

## Automation Tests

- Unit tests: 93 test executions (37 test methods with parameterized data)
- Integration: JKS mode with all 15 algorithm combinations
- Integration: HSM mode with all 15 algorithm combinations

## Security Checks

- Followed secure coding standards
- Ran FindSecurityBugs plugin
- No secrets committed

**Security Note**: This PR implements custom cryptographic code (OAEP padding removal per RFC 3447). Security team review recommended before production deployment.

## Related PRs

**Foundation PR**: #25 (HSM Standard Migration - provides ServerKeystoreRetriever)  
**Documentation PR to Close**: [docs-open-banking #1070](https://github.com/wso2/docs-open-banking/pull/1070) - Documents HSM limitation that this PR resolves. Should be closed/reverted if this custom OAEP implementation is merged.

---

**Files Changed**: 3 files
- JwePayloadDecryptionMediator.java (enhanced with HSM detection)
- HsmJweDecryptionHelper.java (NEW, 343 lines)
- HsmJweDecryptionHelperTest.java (NEW, 370 lines, 20 test methods)

**Lines of Code**:
- Added: ~740 lines (implementation + tests)
- Modified: ~25 lines
- Deleted: ~5 lines

### CODE REVIEW

**Comment by RivinduM** [comment]
shall we change the branch to https://github.com/wso2/financial-services-apim-mediation-policies/tree/hsm_custom_decryption

File: common/jwe-payload-processing/jwe-payload-processor/src/main/java/org/wso2/financial/services/apim/mediation/policies/jwe/processing/util/HsmJweDecryptionHelper.java [hunk 1]
```diff
@@ -0,0 +1,343 @@
+/**
+ * Copyright (c) 2025, WSO2 LLC. (https://www.wso2.com).
```
Comment (RivinduM) [inline common/jwe-payload-processing/jwe-payload-processor/src/main/java/org/wso2/financial/services/apim/mediation/policies/jwe/processing/util/HsmJweDecryptionHelper.java]: update year

Comment (Jenarththan2001) [reply common/jwe-payload-processing/jwe-payload-processor/src/main/java/org/wso2/financial/services/apim/mediation/policies/jwe/processing/util/HsmJweDecryptionHelper.java]: Done! Updated 425b606

**Comment by Jenarththan2001** [comment]
> shall we change the branch to https://github.com/wso2/financial-services-apim-mediation-policies/tree/hsm_custom_decryption

Done! I have updated the base branch of this PR to target hsm_custom_decryption as requested.
